src/nvim/eval.c

```diff
@@ -8495,6 +8495,53 @@ static void f_empty(typval_T *argvars, typval_T *rettv, FunPtr fptr)
   rettv->vval.v_number = n;
 }
 
+/// "environ()" function
+static void f_environ(typval_T *argvars, typval_T *rettv, FunPtr fptr)
+{
+  int i = 0;
+  char_u *entry, *value;
+# ifdef WIN32
+  extern wchar_t **_wenviron;
+# else
+  extern char **environ;
+# endif
+
+  tv_dict_alloc_ret(rettv);
+
+# ifdef WIN32
+  if (*_wenviron == NULL) {
+    return;
+  }
+# else
+  if (*environ == NULL) {
+    return;
+  }
+# endif
+
+  for (i = 0; ; i++) {
+# ifdef WIN32
+    short_u *p;
+
+    if ((p = (short_u *)_wenviron[i]) == NULL) {
+      return;
+    }
+    entry = utf16_to_enc(p, NULL);
+# else
+    if ((entry = (char_u *)environ[i]) == NULL) {
+      return;
+    }
+    entry = vim_strsave(entry);
+# endif
+    if ((value = vim_strchr(entry, '=')) == NULL) {
+      xfree(entry);
+      continue;
+    }
+    *value++ = NUL;
+    tv_dict_add_str(rettv->vval.v_dict, S_LEN((char *)entry), (char *)value);
+    XFREE_CLEAR(entry);
```

Comment:
NOTE: using `xfree` here caused issues (E685: Internal error: hash_add()), Vim uses `vim_free`.  Why is that?

Reply:
No, that is not the issue, still seeing the error - obviously I am using this wrong?